fix: in-scene placed NetworkObject will not spawn client-side when disabled upon being despawned [MTT-4832]

NoelStephensUnity opened the pull request:
If an in-scene placed `NetworkObject`'s `GameObject` was disabled when it was despawned, then when the server re-spawned the same in-scene placed `NetworkObject` the client-side instances would never invoke `NetworkBehaviour.OnNetworkSpawn`. 

[MTT-4832](https://jira.unity3d.com/browse/MTT-4832)
Pertains to #2104

## Changelog
- Fixed: issue where an in-scene placed `NetworkObject` would not invoke `NetworkBehaviour.OnNetworkSpawn` if the `GameObject` was disabled when it was despawned.

## Testing and Documentation
- Includes integration tests.
- Requires updates to documentation ([PR-816](https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/816)).

NoelStephensUnity committed:
fix

This resolves the issue with in-scene placed NetworkObjects that are disabled when despawned not being able to re-spawn again.

NoelStephensUnity committed:
test - manual

Minor adjustments to manually validate the fixes.
Integration test will follow.

NoelStephensUnity committed:
Merge branch 'develop' into fix/in-scene-placed-networkobject-wont-spawn-clientside-when-disabled-upon-despawn-MTT-4832

NoelStephensUnity committed:
fix

This update handles synchronizing despawned in-scene placed NetworkObjects during a scene switch (LoadSceneMode.Single).
Also fixes an exception when no in-scene placed NetworkObject is found.

NoelStephensUnity committed:
test - manual

Cache the NetworkManager to avoid exception during OnDestroy.

NoelStephensUnity committed:
test

Added integration test that validates disabling NetworkObjects when despawned works with currently connected clients, late joining clients, and when scene switching (LoadSceneMode.Single) while also having the server despawn the in-scene placed NetworkObject upon its first spawn (i.e. so it starts off not visible/active to the clients when they finish the scene switch).

NoelStephensUnity committed:
Merge branch 'develop' into fix/in-scene-placed-networkobject-wont-spawn-clientside-when-disabled-upon-despawn-MTT-4832

NoelStephensUnity committed:
update

MTT-4832

NoelStephensUnity committed:
fix

This fixes the issue with the late joining client not being destroyed and removed as well as the issue with not resetting the new properties added to NetworkObjectTestComponent.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -268,7 +268,8 @@ internal void AddSpawnedNetworkObjects()
             m_DespawnedInSceneObjectsSync.Clear();
-            var inSceneNetworkObjects = UnityEngine.Object.FindObjectsOfType<NetworkObject>().Where((c) => c.NetworkManager == m_NetworkManager);
+            // Find all active and non-active in-scene placed NetworkObjects
+            var inSceneNetworkObjects = UnityEngine.Object.FindObjectsOfType<NetworkObject>(true).Where((c) => c.NetworkManager == m_NetworkManager);
```
nit-pick. Can we have
`var inSceneNetworkObjects = UnityEngine.Object.FindObjectsOfType<NetworkObject>(includeInactive: true).Where((c) => c.NetworkManager == m_NetworkManager);` ? Otherwise, it is especially hard to parse what changed in this PR. :-)

jeffreyrainy commented on testproject/Assets/Tests/Manual/SceneTransitioningAdditive/DespawnInSceneNetworkObject.cs:
```diff
@@ -8,87 +8,91 @@ namespace TestProject.ManualTests
+            // despawning.  When re-spawned the client-side will re-activate
+            // the GameObject, while the server-side must set the GameObject
+            // active itself.
+            gameObject.SetActive(false);
```
Are we sure we are not introducing cases where the game developers see their GameObject activated/deactivated outside their code and get confused/unhappy ?

NoelStephensUnity replied:
This particular file is only for manual testing purposes.
The NGO SDK side:
- Never disables any GameObject.
- Only enables a GameObject of an in-scene placed NetworkObject on the client side when the server sends the request to spawn it.

The user has to set the GameObject inactive themselves. If they do this, then when user code wants to spawn the same in-scene placed NetworkObject they just set the GameObject on the server-side to active and then spawn the associated NetworkObject.  The client side NetworkObject clones have to be active in order for OnNetworkSpawn to be invoked, which is what we do only for in-scene placed NetworkObjects.

So, a user should never be surprised that the in-scene placed NetworkObject they disabled gets enabled when their code spawns it from the server side.  

Does this make sense?

NoelStephensUnity committed:
update

adding includeInactive: to FindObjectsOfType where this was changed to true in order to make it easier for reviewers to spot the change.

NoelStephensUnity committed:
Merge branch 'develop' into fix/in-scene-placed-networkobject-wont-spawn-clientside-when-disabled-upon-despawn-MTT-4832

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -388,6 +388,13 @@ internal NetworkObject CreateLocalNetworkObject(NetworkObject.SceneObject sceneO
+                // NetworkBehaviours will have their OnNetworkSpawn method invoked
+                if (networkObject != null && !networkObject.gameObject.activeInHierarchy)
+                {
+                    networkObject.gameObject.SetActive(true);
```
Is this really the behavior we want here? Doesn't this effectively mean users can't disable scene objects at all because we're just going to undo it? Maybe I'm missing context though.

NoelStephensUnity replied:
Look at the comment I left for Jeffrey above, he had the same question.
The short answer is:
This occurs only on the client-side when the server has sent a message to the client to spawn the the disabled NetworkObject.  So, if a user disables a NetworkObject and their code never attempts to spawn it then it will remain disabled.

ShadauxCat replied:
K, sounds fine to me.

Marc477 commented:
With this fix, will it also be possible to Spawn() an object that was previously Despawn() ? So Spawn() and Despawn() at will without destroying the object ? So it keeps running on the server, but not on the client. And only sync with client under certain conditions. I think before there were issues with Spawn() an object that was previously Despawn() but not destroyed

NoelStephensUnity commented:
@Marc477 
For in-scene placed NetworkObjects you are in control of what is disabled and what is not.
So, you could despawn an in-scene placed NetworkObject (passing false to not destroy) and disable it on all clients but on the server you could leave the root GameObject enabled... or you could pick and choose which components you wish to enable/disable like you can normally do in single player. You would just need to make sure to re-enabled whatever is disabled (if anything) on the server side before respawning the same in-scene placed NetworkObject.
